python-checks/src/main/java/org/sonar/python/checks/ExceptionNotThrownCheck.java

```diff
@@ -0,0 +1,83 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.python.checks;
+
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.ClassSymbol;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.semantic.BuiltinSymbols;
+
+// https://jira.sonarsource.com/browse/RSPEC-3984
+// https://jira.sonarsource.com/browse/SONARPY-666
+@Rule(key = "S3984")
+public class ExceptionNotThrownCheck extends PythonSubscriptionCheck {
+  private static final String MESSAGE = "Throw this exception or remove this useless statement.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, check(ExceptionNotThrownCheck::symbolFromInvocation));
+    context.registerSyntaxNodeConsumer(Tree.Kind.NAME, check(ExceptionNotThrownCheck::symbolFromName));
+  }
+
+  private static Consumer<SubscriptionContext> check(Function<Tree, Symbol> extractClassSymbol) {
+    return subscriptionContext -> {
+      Tree t = subscriptionContext.syntaxNode();
+      Symbol symb = extractClassSymbol.apply(t);
+      if (symb != null && symb.is(Symbol.Kind.CLASS) && isThrowable((ClassSymbol) symb)) {
+        Tree parent = t.parent();
+        if (parent.is(Tree.Kind.EXPRESSION_STMT)) {
+          subscriptionContext.addIssue(t, MESSAGE);
+        }
+      }
+    };
+  }
+
+  private static final Set<String> BUILTIN_EXCEPTIONS_FQNS = Stream.of(
+    BuiltinSymbols.EXCEPTIONS,
+    BuiltinSymbols.EXCEPTIONS_PYTHON2).flatMap(Set::stream).collect(Collectors.toSet());
+
+  private static boolean isThrowable(ClassSymbol cs) {
+    return BUILTIN_EXCEPTIONS_FQNS.contains(cs.fullyQualifiedName()) ||
+      cs.superClasses().stream().map(Symbol::fullyQualifiedName).anyMatch(BUILTIN_EXCEPTIONS_FQNS::contains);
```

Comment:
I assume you'd like to use some kind of `ClassSymbol#isOrExtends` here, but you can't because of unresolved type hierarchies for builtin exceptions which would lead to many FNs.
This implementation has also FNs for exception inheriting from custom exceptions:
```
def fn():
  class C1(TypeError): ...
  class C2(C1): ...
  C1() # Noncompliant
  C2() # FN
```

I'm not sure there is currently a way to cover those FNs without risking FPs but I think it would be worth to document the FN (which we could solve later when we cover better the type hierarchies for exception types).

Reply:
Yes, using `isOrExtends` was also my initial idea. I've even added a faithful representation of the exception inheritance hierarchy:

```
private static void addExceptionHierarchy(Map<String, Set<Symbol>> globalSymbols) {
  class Hierarchy {
    final String name;
    final List<Hierarchy> children;
    Hierarchy(String name, Hierarchy... children) {
      this.name = name;
      this.children = Arrays.asList(children);
    }
    public ClassSymbolImpl generate() {
      ClassSymbolImpl classSymbol = new ClassSymbolImpl(name, name);
      for (Hierarchy child: children) {
        ClassSymbolImpl childSymbol = child.generate();
        childSymbol.addSuperClass(classSymbol);
        childSymbol.superClasses();
        globalSymbols.put(child.name, Collections.singleton(childSymbol));
      }
      return classSymbol;
    }
  }
  // https://docs.python.org/3/library/exceptions.html
  Hierarchy throwableHierarchy = new Hierarchy("Throwable",
      new Hierarchy("BaseException",
        new Hierarchy("SystemExit"),
        new Hierarchy("KeyboardInterrupt"),
        new Hierarchy("GeneratorExit"),
        new Hierarchy("Exception",
          new Hierarchy("StopIteration"),
          new Hierarchy("StopAsyncIteration"),
          new Hierarchy("ArithmeticError",
            new Hierarchy("FloatingPointError"),
            new Hierarchy("OverflowError"),
            new Hierarchy("ZeroDivisionError")),
          new Hierarchy("AssertionError"),
          new Hierarchy("AttributeError"),
          new Hierarchy("BufferError"),
          new Hierarchy("EOFError"),
          new Hierarchy("ImportError",
            new Hierarchy("ModuleNotFoundError")),
          new Hierarchy("LookupError",
            new Hierarchy("IndexError"),
            new Hierarchy("KeyError")),
          new Hierarchy("MemoryError"),
          new Hierarchy("NameError",
            new Hierarchy("UnboundLocalError")),
          new Hierarchy("OSError",
            new Hierarchy("BlockingIOError"),
            new Hierarchy("ChildProcessError"),
            new Hierarchy("ConnectionError",
              new Hierarchy("BrokenPipeError"),
              new Hierarchy("ConnectionAbortedError"),
              new Hierarchy("ConnectionRefusedError"),
              new Hierarchy("ConnectionResetError")),
            new Hierarchy("FileExistsError"),
            new Hierarchy("FileNotFoundError"),
            new Hierarchy("InterruptedError"),
            new Hierarchy("IsADirectoryError"),
            new Hierarchy("NotADirectoryError"),
            new Hierarchy("PermissionError"),
            new Hierarchy("ProcessLookupError"),
            new Hierarchy("TimeoutError")),
          new Hierarchy("ReferenceError"),
          new Hierarchy("RuntimeError",
            new Hierarchy("NotImplementedError"),
            new Hierarchy("RecursionError")),
          new Hierarchy("SyntaxError",
            new Hierarchy("IndentationError",
              new Hierarchy("TabError"))),
          new Hierarchy("SystemError"),
          new Hierarchy("TypeError"),
          new Hierarchy("ValueError",
            new Hierarchy("UnicodeError",
              new Hierarchy("UnicodeDecodeError"),
              new Hierarchy("UnicodeEncodeError"),
              new Hierarchy("UnicodeTranslateError"))),
          new Hierarchy("Warning",
            new Hierarchy("DeprecationWarning"),
            new Hierarchy("PendingDeprecationWarning"),
            new Hierarchy("RuntimeWarning"),
            new Hierarchy("SyntaxWarning"),
            new Hierarchy("UserWarning"),
            new Hierarchy("FutureWarning"),
            new Hierarchy("ImportWarning"),
            new Hierarchy("UnicodeWarning"),
            new Hierarchy("BytesWarning"),
            new Hierarchy("ResourceWarning")))));
  globalSymbols.put("Throwable", Collections.singleton(throwableHierarchy.generate()));
}
```

but then discovered that the exceptions are already hardcoded elsewhere, and decided to take the simpler structure that was already there.